lib/Containers/MerkleTree.cpp

```diff
@@ -262,199 +175,224 @@ MerkleTree<Hasher, BranchingBits>::fromBottomMostCompressed(std::string_view buf
-    return tree;
+    return nullptr;
   }
 
   read = slice.get(StaticStrings::RevisionTreeMaxDepth);
   if (!read.isNumber()) {
-    return tree;
+    return nullptr;
   }
-  std::uint64_t maxDepth = read.getNumber<std::uint64_t>();
+  std::uint64_t depth = read.getNumber<std::uint64_t>();
 
   read = slice.get(StaticStrings::RevisionTreeRangeMax);
   if (!read.isString()) {
-    return tree;
+    return nullptr;
   }
   velocypack::ValueLength l;
   char const* p = read.getString(l);
   std::uint64_t rangeMax = basics::HybridLogicalClock::decodeTimeStamp(p, l);
   if (rangeMax == std::numeric_limits<std::uint64_t>::max()) {
-    return tree;
+    return nullptr;
   }
 
   read = slice.get(StaticStrings::RevisionTreeRangeMin);
   if (!read.isString()) {
-    return tree;
+    return nullptr;
   }
   p = read.getString(l);
   std::uint64_t rangeMin = basics::HybridLogicalClock::decodeTimeStamp(p, l);
   if (rangeMin == std::numeric_limits<std::uint64_t>::max()) {
-    return tree;
+    return nullptr;
+  }
+
+  read = slice.get(StaticStrings::RevisionTreeInitialRangeMin);
+  if (!read.isString()) {
+    return nullptr;
+  }
+  p = read.getString(l);
+  std::uint64_t initialRangeMin = basics::HybridLogicalClock::decodeTimeStamp(p, l);
+  if (initialRangeMin == std::numeric_limits<std::uint64_t>::max()) {
+    return nullptr;
+  }
+
+  // summary count
+  read = slice.get(StaticStrings::RevisionTreeCount);
+  if (!read.isNumber()) {
+    return nullptr;
+  }
+  std::uint64_t summaryCount = read.getNumber<std::uint64_t>();
+  
+  // summary hash
+  read = slice.get(StaticStrings::RevisionTreeHash);
+  if (!read.isNumber()) {
+    return nullptr;
   }
+  std::uint64_t summaryHash = read.getNumber<std::uint64_t>();
 
   velocypack::Slice nodes = slice.get(StaticStrings::RevisionTreeNodes);
-  if (!nodes.isArray() || nodes.length() < nodeCountUpToDepth(maxDepth)) {
-    return tree;
+  if (!nodes.isArray() || nodes.length() < nodeCountAtDepth(depth)) {
+    return nullptr;
   }
 
   // allocate the tree
   TRI_ASSERT(rangeMin < rangeMax);
+  auto tree = std::make_unique<MerkleTree<Hasher, BranchingBits>>(depth, rangeMin, rangeMax, initialRangeMin);
 
-  tree = std::make_unique<MerkleTree<Hasher, BranchingBits>>(maxDepth, rangeMin, rangeMax);
-
+  std::uint64_t totalCount = 0;
+  std::uint64_t totalHash = 0;
   std::uint64_t index = 0;
   for (velocypack::Slice nodeSlice : velocypack::ArrayIterator(nodes)) {
     read = nodeSlice.get(StaticStrings::RevisionTreeCount);
     if (!read.isNumber()) {
-      tree.reset();
-      return tree;
+      return nullptr;
     }
     std::uint64_t count = read.getNumber<std::uint64_t>();
 
     read = nodeSlice.get(StaticStrings::RevisionTreeHash);
     if (!read.isString()) {
-      tree.reset();
-      return tree;
+      return nullptr;
     }
     p = nodeSlice.get(StaticStrings::RevisionTreeHash).getString(l);
     std::uint64_t hash = basics::HybridLogicalClock::decodeTimeStamp(p, l);
```

Comment:
Why is the hash encoded as a HLC timestamp? The hash value is obtained by passing the revision ids through some arbitrary hash function, so the hash value can be anything, or am I missing something?

Reply:
Fixed as well.